lib/pg_ha_migrations/safe_statements.rb

```diff
@@ -170,6 +175,99 @@ def safe_remove_concurrent_index(table, options={})
     unsafe_remove_index(table, **options.merge(:algorithm => :concurrently))
   end
 
+  def safe_add_concurrent_partitioned_index(
+    table,
+    columns,
+    name_suffix:,
+    if_not_exists: nil,
+    using: nil,
+    unique: nil,
+    where: nil,
+    comment: nil
+  )
+    raise ArgumentError, "Expected <name_suffix> to be present" unless name_suffix.present?
+
+    if ActiveRecord::Base.connection.postgresql_version < 11_00_00
+      raise PgHaMigrations::InvalidMigrationError, "Concurrent partitioned index creation not supported on Postgres databases before version 11"
+    end
+
+    parent_schema, parent_table = _schema_and_table_for(table)
+
+    fully_qualified_parent_table = "#{quote_schema_name(parent_schema)}.#{parent_table}"
+    parent_index = "#{parent_table}_#{name_suffix}"
+
+    raise PgHaMigrations::InvalidMigrationError, "Table #{parent_table.inspect} is not a partitioned table" unless _partitioned_table?(parent_schema, parent_table)
+
+    _validate_index_name!(parent_index)
+
+    child_schemas_and_tables = select_rows(<<~SQL)
+      SELECT child_ns.nspname, child.relname
+      FROM pg_inherits
+        JOIN pg_class parent        ON pg_inherits.inhparent = parent.oid
+        JOIN pg_class child         ON pg_inherits.inhrelid  = child.oid
+        JOIN pg_namespace parent_ns ON parent.relnamespace = parent_ns.oid
+        JOIN pg_namespace child_ns  ON child.relnamespace = child_ns.oid
+      WHERE parent.relname = #{quote(parent_table)}
+        AND parent_ns.nspname = #{quote(parent_schema)}
+    SQL
+
+    child_tables_with_metadata = child_schemas_and_tables.map do |child_schema, child_table|
+      raise PgHaMigrations::InvalidMigrationError, "Partitioned table #{parent_table.inspect} contains sub-partitions" if _partitioned_table?(child_schema, child_table)
+
+      child_index = "#{child_table}_#{name_suffix}"
+
+      _validate_index_name!(child_index)
+
+      [child_schema, child_table, child_index]
+    end
+
+    # TODO: take out ShareLock after issue #39 is implemented
+    safely_acquire_lock_for_table(fully_qualified_parent_table) do
```

Comment:
If there are blocking queries running against table `foo` in the `public` schema, and you call this method as `safely_acquire_lock_for_table("public.foo") { ... }`, it won't actually match anything here since `tables_with_locks` doesn't contain schema-qualified table names: https://github.com/braintree/pg_ha_migrations/blob/d7ad2e802dae0ffdda62863b495469dd7ec1517d/lib/pg_ha_migrations/safe_statements.rb#L476

Additionally, `safely_acquire_lock_for_table` will not check for locks against partitioned tables since we're only looking at `relkind = 'r'`: https://github.com/braintree/pg_ha_migrations/blob/d7ad2e802dae0ffdda62863b495469dd7ec1517d/lib/pg_ha_migrations/blocking_database_transactions.rb#L59

It _might_ be okay to merge this PR without addressing this issue, but we 100% need to solve this before releasing a new version of the gem (I'm currently working on a fix)